zilliqa/src/consensus.rs

```diff
@@ -566,8 +568,26 @@ impl Consensus {
                 trace!("applying {} transactions to state", transactions.len());
             }
 
+            let transactions: Result<Vec<_>> =
+                transactions.into_iter().map(|tx| tx.verify()).collect();
+            let mut transactions = transactions?;
+
+            // We re-inject any missing Intershard transactions (or really, any missing
+            // transactions) from our mempool. If any txs are unavailable either in the
+            // message or locally, the proposal cannot be applied
+            for (idx, tx_hash) in block.transactions.iter().enumerate() {
+                if transactions.get(idx).is_some_and(|tx| tx.hash == *tx_hash) {
+                    // all good
+                } else {
+                    let Some(local_tx) = self.transaction_pool.pop_transaction(*tx_hash) else {
+                        warn!("Proposal {} at view {} referenced a transaction that was neither included in the broadcast nor found locally - cannot apply block", block.hash(), block.view());
```

Comment:
Not a silly question at all - this is down to the security model of intershard links. Basically intershard messaging is specified to be local-only, i.e. _never p2p_, and as such every validator of a given shard _A_ must have a (light or full) client of any source shards for those shards to be able to send messages to _A_. The validator's sister client will then message the validator node with the `IntershardCall`, populating the mempool.

So a precondition for an intershard tx to be included in a block is that 2/3rds of the validators, including the current leader, have had local clients observe the source call and relay a local message.

This is what allows us to skip all the logic from the UCCB and just accept these transactions as trusted, because we know the validators of _A_ have all locally observed the source transaction, and have verified its authenticity according to the source shard's consensus. Hence why p2p broadcasting is disabled - as p2p messages are inherently untrusted, e.g. a single node would be able to claim an intershard event happened.

Reply:
Maybe it was also James question and if not then it's mine: Is the payload of `Intershard` transactions always included in the block? If the block contained only hashes of `Intershard` transactions, which are not propagated via p2p, then non-validator nodes would never receive their payload and couldn't execute them...

Reply:
Very good point.
Currently, only the hashes are included in Proposals for the current round, but the full body is included in responses to BlockRequest messages. So non-validator nodes would need to request the block explicitly, and wouldn't be able to follow the currently voted-on proposal.

We don't explicitly support non-validator nodes right now but a fix would be for them to simply request the block if they see a Proposal they can't execute. Nodes that want to e.g. run block explorers that want to display the currently voted-on tip, could then spin up sister clients for connected source shard(s) to get the same information that validators do, and everyone else would have to wait until the block vote is accepted and they can get it by requesting from a node.

Reply:
Got it, thanks. Maybe it's worth consireding that all non-validator nodes would request those blocks from the validators because they are the only ones that have the intershard transactions' body due to their sister clients, which might cause unnecessary workload on the validators. Why can't validators actually broadcast intershard transactions to everyone in the network, but rely on intershard transactions they have seen themselves when it comes to voting on a block?